A28-xds-traffic-splitting-and-routing.md

```diff
@@ -0,0 +1,1095 @@
+        will be used,
+        [runtime\_key](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/core/base.proto#L372)
+        will be ignored
+*   Can have
+    [query\_parameters](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L442)
+    *   Note: the entire route containing this will be ignored, regardless of
+        the other matchers, this is the same as this always evaluates to false,
+        because gRPC does not and will not support query parameters)
+*   Must not have
+    [case\_sensitive](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L410)
+    set to false (OK to be unset, default value is true)
+    *   Setting this to false will cause a NACK
+
+The route’s [action
+field](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L181)
+must be
+[route](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L185),
+and its
+[cluster\_specifier](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L696):
+*   Can be
+    [Cluster](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L701)
+*   Can be
+    [Weighted\_clusters](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L719)
+    *   The sum of
+        [weights](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L278)
+        must add up to the
+        [total\_weight](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L335).
+*   Can be
+    [cluster\_header](https://github.com/envoyproxy/envoy/blob/v1.13.1/api/envoy/api/v2/route/route_components.proto#L712)
+    *   Note: the route containing this action will be ignored. This feature can
+        be accomplished by header matching.
+
+Duplicate matches are allowed, but because the first match wins, the second
+will never be used.
+
+Violating any of the validation causes the response to be rejected (NACK’ed).
+The fields not covered by this section will all be ignored.
+
+### Service config generating
+
+The service config generated by the xDS resolver will pick
+`xds_routing_experimental` as the LB policy. As an optimization, when there’s
+only one route, the LB policy picked will be `cds_experimental` or
+`weighted_target_experimental` depending on the action, so we can skip the
+`xds_routing` policy that has only one child.
+
+The `xds_routing_experimental` balancer config will contain the validated
+`{matchers, action}` pairs. The actions will be in a separate list, and will
+be referred to by the routes by their names.
+
+#### Action list
+
+All the actions from the validated `{matchers, action}` pairs will be
+de-duplicated and stored in a cache. All actions in the cache will be
+assigned a unique name.
```

Comment:
Is that a global cache or cache per RouteConfiguration? Is the unique name internally generated or does it map to anything in the xDS response?

Reply:
Global, between RDS responses.